### PR TITLE
Create Convert to data API for ContigSet and Genome

### PR DESCRIPTION
**Pull request by cshenry**
This is the PD for the transition of methods to data API discussed at EC meeting.

### CODE REVIEW

**Comment by scanon** [comment]
@cshenry  Can you comment on which of those methods would benefit more from a Data-to-files method that understands the new objects?

**Comment by cshenry** [comment]
Not the modeling. Probably not genome annotation. The other modules I know too little of to comment on. I think it will be true for some... probably many. 

I was thinking that one way we might handle this is to make a data-to-files method that creates the old genome object... I know the data-API guys already did this, and I think we can use their code... but I'm a bit concerned that their code actually caches the object in the workspace. I'm wondering why they did this... just how long does it take to make an old genome object for a microbial genome using the data API?

**Comment by scanon** [comment]
@cshenry   Can you rename this to .md so it renders correctly?

**Comment by vkcshl** [comment]
@cshenry kb_cummerbund should be moved to the same group as KBaseRNASeq module. It is being refactored to use Data API since some of the table widgets display functional information for selected genes (e.g. genes selected from volcano plot). This is scheduled for completion before Q3.

**Comment by mlhenderson** [comment]
Many objects contain Genome references only, and for those references to be
valid they must have a Genome object.  You cannot substitute another type
without generating an error on save or changing all the types.

If there are some optimizations to make, for instance if no code uses
ContigSet, then we can simply make a Genome without contigs.

Bottom line is that our system will only work as expected by existing code
if objects are saved.  Unless we are massively overhauling code and types
before the end of Q2, which nobody is working on.

This is not caching, although it does in fact take time to pull the data
from workspace for the new types because there is more information
captured, then dump to the old format.

**Comment by cshenry** [comment]
But this means all these objects are pointing to some copy of the old genome object rather than the actual genome object, which seems counter to the concept of provenance etc.

It seems like a simple solution would be if Workspace would accept refs that have no associated type… or could be one of two types (e.g. old Genome, or new GenomeAnnotation). This isn’t ideal either, but still seems better than pointing to a copy of the genome rather than  the actual genome.

Another route is to keep two refs in these objects (both optional). One would point to an old genome object and one to the new. Again, not an ideal solution, but this last solution probably does the best possible job of preserving provenance… Refactoring objects in this way wouldn’t be that difficult...

On May 25, 2016 at 12:35:32 PM, Matt Henderson (notifications@github.com) wrote:

Many objects contain Genome references only, and for those references to be
valid they must have a Genome object. You cannot substitute another type
without generating an error on save or changing all the types.

If there are some optimizations to make, for instance if no code uses
ContigSet, then we can simply make a Genome without contigs.

Bottom line is that our system will only work as expected by existing code
if objects are saved. Unless we are massively overhauling code and types
before the end of Q2, which nobody is working on.

This is not caching, although it does in fact take time to pull the data
from workspace for the new types because there is more information
captured, then dump to the old format.

**Comment by MrCreosote** [comment]
> It seems like a simple solution would be if Workspace would accept refs that have no associated type… or could be one of two types (e.g. old Genome, or new GenomeAnnotation). This isn’t ideal either, but still seems better than pointing to a copy of the genome rather than  the actual genome.

It can do that: https://kbase.us/services/ws/docs/typedobjects.html#idannotations

**Comment by mlhenderson** [comment]
How many things actually modify the Genome object?  Only RAST annotation
that I am aware of.

There are zero things currently that modify the GenomeAnnotation object, so
there would be no provenance to save for the Genome object derived from it.
We would save in the Genome provenance that it came from the
GenomeAnnotation object, and that should then cover the provenance
collection.

If other entities are a consumer only for a Genome, then they record their
own provenance.

We certainly can modify all types in the system that currently refer to a
Genome object, but that then requires all code that connects to those
objects to also change, be tested, etc.  This includes any Apps/services,
widgets, landing pages, etc that consume those types.  I don't think this
is a trivial task.  Who would do this work before the end of Q2?

**Comment by msneddon** [comment]
Maybe I'm missing something here with the provenance and references between objects, or I don't understand our current transition plan, but I think if the data API has methods for accessing/saving the new genome type in a way that looks like the old genome type, then we would have a clear path to migrating to the new genome type that sidesteps some of these issues.  I could still be wrong, but I don't think we would have to duplicate objects in the workspace in the old and new formats....

I doubt this could all be completed in this quarter, but a transition plan could look something like:

0) Current methods interact with the old Genome/ContigSet object through the workspace, and there are only a small number of methods that can possibly be used: ws.get_objects(..), ws.get_object_subset(..), and ws.save_objects().

1) Make sure the data API can reproduce these methods:
api.get_legacy_genome/contigset(...), api.get_legacy_genome/contigset_subset(..) and api.save_legacy_genome/contigset(..)

The critical bit here is that 'get_legacy' methods would work for both the new and old types, and the 'save_legacy' could be configured in the data api to save either the old or new genome type (so we can decide precisely when to cut over to creating the new type).

2) Update all the Genome object reference fields in the workspace types so they can point to either the old or new genome objects.  This change alone would not break anything and could be done in a few hours by searching the typespecs registered in the workspace.

3) Switch over existing App implementations and landing pages to use this API.  Really, only the API calls would change- all the existing code would continue to work because you can still get/save genome objects in the old format without touching any code logic.  This may still be a lot of work, but is straightforward I think to just systematically look at every get/save call to the workspace for genomes and contigs.  Importantly, we can update apps/landing pages one at a time because the new API methods should work just fine with the existing genome type.

4) For every App implementation that is updated, switch over Narrative method spec to accept both the old type and new Genome type.  This is a simple method specification change that would be visible as soon as we want.  The current methods would still populate lists correctly and behave the same.

At this point, users still upload the old genome type, the tools still operate on the old type, and when new versions are saved through the API, it still creates the old type.  But, everything is now going through the API.

5) When we think we are ready, we throw the switch so that the API starts saving genomes in the new format.  We would slowly begin to accumulate new genome types in the system, and we can make sure that everything works cleanly with them.  If things start to break, we can quickly revert the API service and fix things and try again (or do this on CI/next...).  Note that in production, users are still uploading the old genome type.  There is still only ever one object per genome in the system.

6) Finally, when we are comfortable, we make a clean cut by switching the uploader to the new Genome type, and completely remove the current uploader.  Then all new data in the system would come in as the new Genome object without the need for creating the old object.  Users would not have to select between the two uploaders or have to decide which genome works with which App.

What does everyone think?  Am I missing something important?

**Comment by mlhenderson** [comment]
The Data API was not designed to work in the way that you are describing,
and this is in fact counter to the goals of Data API, which is to move away
from knowledge or direct dependence on the object schemas in code outside
of the API.  That fundamental aspect of schema reliance is crippling the
system from being able to easily handle type changes, and this is a problem
that Data API is aimed at.

The current transition plan is as follows:

Q2 - Enable upload of GenomeAnnotation and Assembly, which includes a
converted Genome and ContigSet for anything in the system that does not
understand GenomeAnnotation or Assembly.  Disable upload of Genome and
ContigSet so that there is only one upload type, which will help reduce
confusion and enforce data consistency for uploaded data.  Current code in
the system that relies on Genome or ContigSet does not need to change at
all, and all Apps that use Genome or ContigSet as input will continue to
work as expected.  RNA-seq will accept GenomeAnnotation and Assembly
objects, and users can upload Eukaryotes for RNA-seq.  This builds plenty
of time for other code in the system to transition, and can be done in
stages while reaping all of the benefits of the new types for users of
Eukaryotic data, while losing no functionality for users of Microbial data.

Q3 - Begin transition of other code in the system that depends on Genome or
ContigSet.  This includes updating types to use the new object type
references, modifying code to transition from the old object types to Data
API, including Apps, widgets, landing pages.  This should happen for the
most critical pieces first from end to end, then fan out to less critical
pieces.  This seems to be the least disruptive path while still allowing
updates in Q2 to proceed to production.

This fits with the current scope of work that is happening in Q2 and seems
to be scheduled for Q3.

**Comment by msneddon** [comment]
I guess I see the immediate data problem right now as the tight coupling between how we store data and the interface to that data, and not so much how an application operates on data once it has it.  As we know, if we change the structure of data in the Workspace, it changes immediately the interface to that data, so things break unless all code using that data is updated.

If there is a software layer above the workspace for specific data types, then we can change how we store that data at any time, and as long as we maintain the interface to that middle layer, then everything continues to work.  I think we all agree on that.  Once an application gets that data in some particular format though, I don't think we can or should try to enforce how that data is structured in memory or operated on.

So I guess this is all just to say that I believe we will break our schema reliance problem only once all applications use a data API layer rather than directly fetching the raw data.  And if that is our goal, I don't see why having a method in that data API or a data<->files method that returns data in a legacy format would be an issue.  It's just an interface to data at that point, even if it is a poor interface, but gives us a way to switch to that interface quickly.

It's not my call, but my personal opinion is that having two different representations of the same uploaded data visible to users and developers will be incredibly confusing without substantial updates to the Narrative Data panel and input widgets, especially if some apps work on one representation of the data but not others.  I understand that that would be temporary, but it seems like until we refactor nearly everything, developers will still have to think about what format the genome is stored in when dealing with any data that references a genome.

**Comment by mlhenderson** [comment]
Of course I share your concerns about confusing users.  I think with good
labeling and documentation we can minimize that during the transition
period, and it will be clear which Apps have inputs for Genome vs
GenomeAnnotation or ContigSet vs Assembly.

In order to be able to evolve the types in our system, we will definitely
need to have a way for more than one type to exist at the same time, even
if only to transition from one to the other, but sometimes it may be
necessary for similar entities to exist until they can be resolved.  As the
system grows, it will become increasingly long and difficult to change out
one type for another in full as is being suggested.  It is already long and
difficult now, although my plan is to drastically improve that going
forward.

Switching from direct schema calls to Data API calls is really pretty
straightforward.  I would like people to at least try to use the methods
that are there before deciding that they just need the old JSON in full.
I'm concerned about the obfuscation of code if refactoring to use a handful
of methods is significantly harder than asking for the entire object JSON
at once.  We've had to eat our own dogfood with the API and actually use
it, and I am saying it is not that difficult.  Could it be improved? Sure,
it is not perfect.

We are basically forced as of today to make the Genome and ContigSet
objects as well, because of existing references.  Of course it is possible
to update this, although the chain effect of doing this is definitely a
non-trivial amount of work overall and requires full system testing.  This
is not going to happen in Q2, and I don't have the resources to do all of
it in Q3, it will take multiple teams to cover the full scope I think.  The
references issue will also continue to be a problem going forward for other
types, but it may be more localized than Genome, remains to be seen when
the time comes.

We do have an ObjectAPI that provides a wrapper interface to workspace for
fetching the object JSON directly of an object in use.  If it turned out
that it was very difficult to use the existing API methods (which it would
be helpful to know exactly where and why that is so that we can either make
improvements to the API or understand what makes the code so dependent on
the schema), then we can store a pointer from the new type to the old type
and a method that would return the reference so you could then access it as
you would have before, either by requesting the entire object or requesting
a subset via paths.  So this isn't exactly what you're asking for, but it
is very close and is generic.  You can use the ObjectAPI for any type,
including the Genome types, but you don't have any of the high level
interfaces like the rest of the Data API, just basic object methods.  The
other APIs are built using this as the core.

We still can't stop the creation of the Genome and ContigSet objects until
all the references are resolved.  That does not seem like it is going to
happen in full in Q2, and possibly not in full in Q3.  Modifying the type
specs is not much work by itself, but updating all the places where that
data is created and used and then testing the system in full is definitely
a non-trivial amount of work.  I think it is very likely that this will
happen in stages and not in one sweep with a transition to Data API, but if
it can happen in one sweep that would be great.  At that point, once all
references were updated and all code was updated, we could turn off
creation of Genome and ContigSet.  I see that taking a while though.

Anyway, it takes much less than a day for a developer to look at the API
docs and look at their code to evaluate how much updating there would be.
